src/triage/predictlist/__init__.py

```diff
@@ -0,0 +1,593 @@
+            total_nonimpute_cols = set(f for f in set(feature_names) if '_imp' not in f) - total_impute_cols
+            
+            task_generator = self.feature_generator._generate_imp_table_tasks_for
+            
+            imputation_table_tasks.update(task_generator(
+                aggregation,
+                impute_cols=list(total_impute_cols),
+                nonimpute_cols=list(total_nonimpute_cols)
+                )
+            )
+        return reconstructed_feature_dict, imputation_table_tasks
+    
+    def retrain(self, today):
+        """Retrain a model by going back one split from today, so the as_of_date for training would be (today - training_label_timespan)
+        
+        Args:
+            today (str) 
+        """
+        today = dt_from_str(today)
+        as_of_date = datetime.strftime(today - convert_str_to_relativedelta(self.training_label_timespan), "%Y-%m-%d")
+ 
+        retrain_definition = {
+            'first_as_of_time': dt_from_str(as_of_date),
+            'last_as_of_time': dt_from_str(as_of_date),
+            'matrix_info_end_time': today,
+            'as_of_times': [dt_from_str(as_of_date)],
+            'training_label_timespan': self.training_label_timespan,
+            'training_as_of_date_frequency': self.experiment_config['temporal_config']['training_as_of_date_frequencies'],
+            'max_training_history': self.experiment_config['temporal_config']['max_training_histories'][0],
+        }
+        cohort_table_name = f"triage_production.cohort_{self.experiment_config['cohort_config']['name']}_retrain"
+ 
+        # 1. Generate all labels
+        self.generate_all_labels(as_of_date)
+
+        # 2. Generate cohort
+        self.generate_entity_date_table(as_of_date, cohort_table_name)
+
+        # 3. Generate feature aggregations
+        collate_aggregations = self.get_collate_aggregations(as_of_date, cohort_table_name)
+        feature_aggregation_table_tasks = self.feature_generator.generate_all_table_tasks(
+            collate_aggregations,
+            task_type='aggregation'
+        )
+        self.feature_generator.process_table_tasks(feature_aggregation_table_tasks)
+
+        # 4. Reconstruct feature disctionary from feature_names and generate imputation
+        feature_imputation_table_tasks = self.feature_generator.generate_all_table_tasks(
+            collate_aggregations,
+            task_type='imputation'
+        )
+        self.feature_generator.process_table_tasks(feature_imputation_table_tasks)
+        
+        feature_dict = self.feature_dictionary_creator.feature_dictionary(
+            feature_table_names=feature_imputation_table_tasks.keys(),
+            index_column_lookup=self.feature_generator.index_column_lookup(collate_aggregations),
+        )
+        feature_group_creator = FeatureGroupCreator({"all": [True]})
+        feature_group_mixer = FeatureGroupMixer(["all"])
```

Comment:
I haven't done this code diving in a while, so I have to ask why the group creator and mixer can be 'all'. What happens if this model happened to be created with one of the leave-one-in groups? Would the base feature dictionary already be subsetted?

If this is the case, then I *think* you can actually ignore the FeatureGroupCreator/Mixer and pass the initial feature_dict directly to make_metadata and build_matrix.

Reply:
yeah I see the point. The reason I put it "all" is because features are not locked in with the given `model_group_id`. So if we are going to use the base feature dictionary, does that mean that we are using the feature dictionary of the last model from the given model_group_id?

Reply:
I think `feature_groups` should be specified at the `model_group_id` level, shouldn't they? In any case, it's been a while since I've really wrapped my mind around the feature groups code as well, but we should probably at least include a test here to check that the right subset of features is getting used to train the new model if the original was part of a leave-one-in/out experiment.

Reply:
I think the features are locked in with the model group. The `feature_list` saved with it (https://github.com/dssg/triage/blob/master/src/triage/component/results_schema/schema.py#L128) should be the final feature list, post-grouping and mixing. This enables audition to work; if you want to compare models of different feature groups against each other, you need them in different model groups as Audition works at the model group level.